### ISSUE TITLE
Clarification on the transfer process / improve status messages

### ISSUE THREAD

**Issue by w1th0utnam3** [schollz/croc#443]
When trying to send a folder containing many files, especially if the files are big, it takes quite some time before `croc` shows the  passphrase and it looks like nothing is happening (the output "Sending 0 files (X.XX GB)" **after** the first file was processed is also not very descriptive). It is not clear what happens during that stage. At first I thought it was only hashing the files but task manager already shows quite some network traffic? It would be nice if the readme had a section that actually describes on a high level what `croc` is doing in each stage during the transfer process. Your blog post describes the basic ideas but not what the actual workflow is. Currently unclear to me is for example:

1. As indicated by my introduction: what is happening in the stage before `croc` outputs the passphrase? Is it only hashing? Does it already transfer data to the relay?
2. Does the relay have to process all the traffic of the data transfer or is it only needed to set up a direct tunnel between sender and receiver?

I think these are some basic information that should be made clear in the readme. Unfortunately I'm not familiar enough with how `croc` works to contribute this.

Maybe a separate issue, but related, is that it would be nice if `croc` showed a message like "Computing hash of file ? (file #X of Y)" during this "pre-processing" stage before you receive the passphrase so that the user knows that `croc` is doing *something*.

**Comment by schollz** [comment]
Before the passphrase, croc is hashing the files. The croc outputs this information now.

The relay passes the encrypted data between the sender and receiver.  I'm curious to what is meant by "direct tunnel"? The relay is a fixed point along the data's journey, but afaik, unless you are on LAN, there is [never a "direct tunnel"](https://gsuite.tools/traceroute).

**Comment by w1th0utnam3** [comment]
Thanks for the reply. I'm not very familiar with networking in general so "direct tunnel" is probably bad wording so let me rephrase it. AFAIK depending on the network environment of the two clients it is in principle possible to establish a connection for data transfer without a server that relays every byte of the transfer. For example in the WebRTC stack a STUN server helps the clients to automatically identify possible IP address + port pairs that could allow a direct connection without the user manually having to specify any addresses (e.g. when non-static addresses are involved). Only if this doesn't work out (e.g. because of firewall settings?) I think WebRTC usually falls back to a TURN server which would relay all traffic between the clients. 

**Regarding my initial post**

My point was that `croc` supports a direct connection via LAN (without relay) and otherwise always transfers via the relay. IMO it could be formulated a bit more clearly in the readme for people that don't know anything about networking that this is how the tool operates. The phrase "allows any two computers to transfer data (using a relay)" of course hints at this but I think that it's not completely clear what this means. In my opinion adding a sentence explaining that the official relay server (hosted with DigitalOcean) has to repeat every byte would make sense.

**Additional thoughts**

Because this is the way the tool operates, I'm a bit hesitant to use it to transfer e.g. files of several GBs because this wastes a lot of bandwidth of the official relay. Going back to the first paragraph, I saw a discussion in an issue about using [WebRTC for the relay](https://github.com/schollz/croc/issues/108). Slightly related to the most recent comments there, maybe you could consider adopting only the STUN part of the stack, i.e. to allow establishing a direct connection without the need for a relay server. I played a bit around with a self-written prototype that uses a public STUN server combined with manually sending the found connection candidates to a friend via an external chat program. To my surprise I was still able to directly transfer data to my friend using these connection candidates even though we are not in a LAN, both behind routers with standard firewall setting, while our (cable) internet providers use some DSlite technology + sharing of IPv4 address over multiple customers which apparently makes operating an IPv4 web server nearly impossible (just FYI to explain why it surprised me that it worked). So I think this approach could work in surprisingly many cases. Combined with an optional command line flag to only allow connections without a relay this could help considerate users avoid congesting the relay server.

**Comment by schollz** [comment]
> IMO it could be formulated a bit more clearly in the readme

Would appreciate a PR for how you would like the clarification

> WebRTC usually falls back to a TURN server

In my experience this is quite frequently the case to fallback to TURN. And then its the same as croc (transferring with a relay). I don't like the idea of using "public" TURN servers for anything because there's no knowledge of their uptime, throttling, etc. But, suppose I did develop croc with STUN/TURN...is it possible to make my own STUN/TURN servers that can't be co-opted by any other WebRTC process (my thinking is that they are all the same protocol, so they could...)?

**Comment by schollz** [comment]
This is a reasonable feature request. Would you mind making a new issue for it so it can be tracked easier?
